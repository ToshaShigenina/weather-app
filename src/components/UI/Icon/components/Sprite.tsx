import React, { FC, SVGAttributes, useEffect } from 'react';
import { createPortal } from 'react-dom';
import { ReactComponent as SpriteSvg } from '../svg/sprite.svg'

import '../css/sprite.css';

const portalNode = document.createElement('div');
portalNode.setAttribute('id', 'sprite');
portalNode.setAttribute('data-testid', 'sprite');
portalNode.hidden = true;

export const Sprite: FC<SVGAttributes<SVGElement>> = (props) => {
	useEffect(() => {
		document.body.appendChild(portalNode);
		return () => portalNode.remove();
	});
	return createPortal(<SpriteSvg { ...props } />, portalNode)
};
