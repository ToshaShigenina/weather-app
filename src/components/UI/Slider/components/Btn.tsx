import React, { ButtonHTMLAttributes, FC } from 'react';
import cn from 'classnames';
import { ReactComponent as SliderArrow } from '../svg/slider-arrow.svg'

type SliderBtnDirectionType = 'prev' | 'next';
interface SliderBtnProps extends ButtonHTMLAttributes<HTMLButtonElement> {
    direction: SliderBtnDirectionType
}

export const Btn: FC<SliderBtnProps> = ({ direction = '', className, ...otherProps }) => {
    return (
        <button
            className={ cn(className, 'slider__btn', direction && `slider__btn-${direction}`) }
            { ...otherProps }
        >
            <SliderArrow />
        </button>
    );
};
