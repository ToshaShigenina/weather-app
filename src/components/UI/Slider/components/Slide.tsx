import React, { FC, HTMLAttributes, useContext, useEffect, useState } from 'react';
import cn from 'classnames';
import { SliderContext, type SliderContextType } from '../context';

export const Slide: FC<HTMLAttributes<HTMLDivElement>> = ({ children, className = '', style = {}, ...otherProps }) => {
    const slideKey = Symbol();
    const [slideIndex, setSlideIndex] = useState(0);
    const sliderCtx = useContext<SliderContextType | null>(SliderContext);
    const width = sliderCtx && sliderCtx.width > 0
        ? sliderCtx.width + 'px'
        : '100%';
    const marginLeft = sliderCtx?.margin || 0;

    useEffect(() => {
        if (sliderCtx && sliderCtx.addSliderKey && !sliderCtx.sliderKeys.includes(slideKey)) {
            setSlideIndex(sliderCtx.sliderKeys.length);
            sliderCtx.addSliderKey(slideKey);
        }
    }, []);
    return (
        <div
            className={ cn(className, 'slide') }
            style={ {
                ...style,
                width,
                marginLeft: slideIndex ? marginLeft : 0,
            } }
            { ...otherProps }
        >
            { children }
        </div>
    );
};
