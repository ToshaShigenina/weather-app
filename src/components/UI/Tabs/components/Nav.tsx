import React, { FC, OlHTMLAttributes } from "react";
import cn from "classnames";

export interface NavProps extends Omit<OlHTMLAttributes<HTMLUListElement>, 'children' | 'onClick'> {
	active?: number,
	items?: string[],
	onClick?: (index: number) => void
}

export const Nav: FC<NavProps> = ({ active = 0, items = [], onClick, ...othrProps }) => {
	const onClickHandler = (index: number) => {
		if (onClick) {
			onClick(index);
		}
	};

	return (
		<>
			{
				Boolean(items.length) &&
				<ul className={ 'tabs-nav' } { ...othrProps }>
					{
						items.map((item, i) => {
							return (
								<li className={ cn('tabs-nav__item', active === i && 'active') } key={ 'tabs-nav-item-' + i }>
									<button
										className="tabs-nav__btn"
										onClick={ () => onClickHandler(i) }
									>
										{ item }
									</button>
								</li>
							)
						})
					}
				</ul>
			}
		</>
	);
};
