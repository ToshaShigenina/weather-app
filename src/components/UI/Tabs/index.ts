export { Tabs } from './components/Tabs';
export type { TabItem, TabsProps } from './components/Tabs';