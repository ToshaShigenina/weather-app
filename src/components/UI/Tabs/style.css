.tabs-nav {
	display: flex;
	justify-content: center;
	gap: 0 50px;
	list-style: none;
	margin-bottom: 30px;
}

.tabs-nav__btn {
	--color: var(--color-main-light);
	--color-border: transparent;
	display: block;
	flex-shrink: 0;
	border: none;
	border-radius: 0;
	background-color: transparent;
	padding-bottom: 10px;
	border-bottom: 4px solid var(--color-border);
	color: var(--color);
	font-weight: 700;
	font-size: 20px;
	line-height: 1;
	white-space: nowrap;
	cursor: pointer;
}

.tabs-nav__item.active .tabs-nav__btn {
	--color: var(--color-main-dark);
	--color-border: var(--color-main-dark);
	cursor: default;
}

@media screen and (min-width: 768px) {
	.tabs-content {
		max-width: 224px;
		margin: 0 auto;
	}

	.tabs-content .slider__btn-prev {
		left: -52px;
	}

	.tabs-content .slider__btn-next {
		right: -52px;
	}
}

@media screen and (min-width: 912px) {
	.tabs-content {
		max-width: 348px;
	}

	.tabs-nav {
		gap: 0 16px;
	}
}

@media screen and (min-width: 1024px) {
	.tabs-content {
		max-width: 472px;
	}
}

@media screen and (min-width: 1200px) {
	.tabs-content {
		max-width: 596px;
	}
}

@media screen and (min-width: 1360px) {
	.tabs-content {
		max-width: 720px;
	}
}