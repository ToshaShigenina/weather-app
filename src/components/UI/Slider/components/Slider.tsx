import React, { FC, HTMLAttributes, useRef, useState, useEffect } from 'react';
import { Btn } from './Btn';
import { useResize } from '../../../../hooks';
import { getCountSlides, getMinBreackpoint, getBreackpointsMap } from '../utils'
import { SliderContextProvider, type SliderContextType } from '../context';
import '../style.css';

export type SliderBreackpointProps = {
    slidesPerView?: number,
    marginBetweenSlides?: number,
}
export interface SliderProps extends HTMLAttributes<HTMLDivElement>, SliderBreackpointProps {
    breackpoints?: Record<number, SliderBreackpointProps>,
}

export const Slider: FC<SliderProps> = ({ children = null, slidesPerView = 1, marginBetweenSlides = 0, breackpoints, ...otherProps }) => {
    const sliderWrapperRef = useRef(null);
    const { width: windowWidth } = useResize();
    const [currentSlide, setCurrentSlide] = useState(1);
    const [slideWidth, setSlideWidth] = useState(0);
    const [countSlides, setCountSlides] = useState(0);
    const [countSlidesPerView, setCountSlidesPerView] = useState(slidesPerView);
    const [sliderGap, setSliderGap] = useState(marginBetweenSlides);
    const [translate, setTranslate] = useState(0);
    const [sliderKeys, setSliderKeys] = useState<symbol[]>([]);
    const context: SliderContextType = {
        width: slideWidth,
        margin: sliderGap,
        sliderKeys,
        addSliderKey: (key: symbol) => {
            sliderKeys.push(key);
            setSliderKeys(sliderKeys);
        }
    };

    const calcSlideWidth = () => {
        const count = getCountSlides(children);
        setCountSlides(count);
        if (count > 0 && sliderWrapperRef && sliderWrapperRef.current) {
            const sliderWidth = (sliderWrapperRef.current as HTMLDivElement).clientWidth
            setSlideWidth((sliderWidth - ((countSlidesPerView - 1) * sliderGap)) / countSlidesPerView);
        }
    };
    useEffect(() => {
        if (breackpoints) {
            const bpMap = getBreackpointsMap(breackpoints, windowWidth);
            if (Object.values(bpMap).every((item) => !item)) {
                setCountSlidesPerView(slidesPerView);
                setSliderGap(marginBetweenSlides);
            } else {
                const bpMin = getMinBreackpoint(bpMap);
                const bpTemp = breackpoints[bpMin];
                if (bpTemp.slidesPerView !== undefined) setCountSlidesPerView(bpTemp.slidesPerView);
                if (bpTemp.marginBetweenSlides !== undefined) setSliderGap(bpTemp.marginBetweenSlides);
            }
        }
        calcSlideWidth();
    }, [windowWidth]);
    useEffect(() => {
        calcSlideWidth();
    }, [children, countSlidesPerView, sliderGap]);

    const clickToPrevBtn = () => {
        if (currentSlide > 1) {
            setCurrentSlide(currentSlide - 1)
            setTranslate(translate + (slideWidth + sliderGap));
        }
    };
    const clickToNextBtn = () => {
        if (currentSlide <= countSlides - countSlidesPerView) {
            setCurrentSlide(currentSlide + 1)
            setTranslate(translate - (slideWidth + sliderGap));
        }
    };

    return (
        <SliderContextProvider value={ context }>
            <div className="slider" { ...otherProps }>
                <div className="slider__wrapper" ref={ sliderWrapperRef }>
                    <div
                        className="slider__container"
                        style={ { transform: `translateX(${translate}px)` } }
                    >
                        { children }
                    </div>
                </div>
                <Btn
                    onClick={ clickToPrevBtn }
                    direction="prev"
                    disabled={ currentSlide <= 1 }
                />
                <Btn
                    onClick={ clickToNextBtn }
                    direction="next"
                    disabled={ currentSlide > countSlides - countSlidesPerView }
                />
            </div>
        </SliderContextProvider>
    );
};
