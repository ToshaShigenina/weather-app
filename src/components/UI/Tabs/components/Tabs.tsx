import React, { FC, HTMLAttributes, ReactNode, useState } from "react";
import cn from "classnames";
import { Nav } from './Nav';
import '../style.css';

export type TabItem = {
	label: string
	children: ReactNode
};
export interface TabsProps extends Omit<HTMLAttributes<HTMLDivElement>, 'children'> {
	items: TabItem[]
};
type LabelsFunc = (items: TabItem[]) => string[];

const getLabels: LabelsFunc = (items) => {
	return items.map(item => item.label);
};
export const Tabs: FC<TabsProps> = ({ items = [] }) => {
	const [active, setActive] = useState(0);

	const changeActiveTab = (index: number) => {
		setActive(index);
	};

	return (
		<>
			<Nav active={ active } items={ getLabels(items) } onClick={ changeActiveTab } />
			<div className="tabs-content">
				{ items[active].children }
			</div>
		</>
	);
};
