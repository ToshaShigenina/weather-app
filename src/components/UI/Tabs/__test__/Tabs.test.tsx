import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { mountTest } from '../../../../tests';
import { Tabs, type TabsProps } from '../components/Tabs';

const testId = 'tabs';

describe('Tabs', () => {
	mountTest<TabsProps>(Tabs, {
		items: [
			{
				label: '',
				children: ''
			}
		]
	});

	it('компонент рендерится с двумя кнопками и контентом, активна первая кнопка и отображается первый контент', () => {
		const items = [
			{
				label: 'Item 1',
				children: 'Children 1',
			},
			{
				label: 'Item 2',
				children: 'Children 2',
			},
		];
		render(<Tabs items={ items } data-testid={ testId } />);
		expect(screen.getByTestId(testId)).toBeInTheDocument();
		const navItems = screen.getByTestId(testId).querySelectorAll('li');
		const content = screen.getByTestId(testId).querySelector('.tabs__content');
		expect(navItems.length).toEqual(items.length);
		expect(content).not.toEqual(null);
		expect(navItems[0]).toHaveClass('active');
		expect(navItems[0].textContent).toContain(items[0].label);
		expect(navItems[1]).not.toHaveClass('active');
		expect(navItems[1].textContent).toContain(items[1].label);
		content && expect(content.textContent).toContain(items[0].children);
		expect(screen.getByTestId(testId)).toMatchSnapshot();
	});

	it('компонент рендерится с двумя кнопками и контентом, при клике на кнопку происходит изменение контента', () => {
		const items = [
			{
				label: 'Item 1',
				children: 'Children 1',
			},
			{
				label: 'Item 2',
				children: 'Children 2',
			},
		];
		render(<Tabs items={ items } data-testid={ testId } />);
		expect(screen.getByTestId(testId)).toBeInTheDocument();
		const navItems = screen.getByTestId(testId).querySelectorAll('li');
		const content = screen.getByTestId(testId).querySelector('.tabs__content');
		expect(navItems.length).toEqual(items.length);
		expect(content).not.toEqual(null);
		expect(navItems[0]).toHaveClass('active');
		expect(navItems[0].textContent).toContain(items[0].label);
		expect(navItems[1]).not.toHaveClass('active');
		expect(navItems[1].textContent).toContain(items[1].label);
		content && expect(content.textContent).toContain(items[0].children);
		expect(screen.getByTestId(testId)).toMatchSnapshot();
		const btn = navItems[1].querySelector('button');
		expect(btn).not.toEqual(null);
		btn && fireEvent.click(btn);
		expect(navItems[0]).not.toHaveClass('active');
		expect(navItems[1]).toHaveClass('active');
		content && expect(content.textContent).toContain(items[1].children);
		expect(screen.getByTestId(testId)).toMatchSnapshot();
	});
})