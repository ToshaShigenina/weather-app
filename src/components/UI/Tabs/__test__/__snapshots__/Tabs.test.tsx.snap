// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs компонент рендерится с двумя кнопками и контентом, активна первая кнопка и отображается первый контент 1`] = `
<body>
  <div>
    <ul
      class="tabs-nav"
    >
      <li
        class="tabs-nav__item active"
      >
        <button
          class="tabs-nav__btn"
        >
          Item 1
        </button>
      </li>
      <li
        class="tabs-nav__item"
      >
        <button
          class="tabs-nav__btn"
        >
          Item 2
        </button>
      </li>
    </ul>
    <div
      class="tabs-content"
      data-testid="tabs"
    >
      Children 1
    </div>
  </div>
</body>
`;

exports[`Tabs компонент рендерится с двумя кнопками и контентом, при клике на кнопку происходит изменение контента 1`] = `
<body>
  <div>
    <ul
      class="tabs-nav"
    >
      <li
        class="tabs-nav__item active"
      >
        <button
          class="tabs-nav__btn"
        >
          Item 1
        </button>
      </li>
      <li
        class="tabs-nav__item"
      >
        <button
          class="tabs-nav__btn"
        >
          Item 2
        </button>
      </li>
    </ul>
    <div
      class="tabs-content"
      data-testid="tabs"
    >
      Children 1
    </div>
  </div>
</body>
`;

exports[`Tabs компонент рендерится с двумя кнопками и контентом, при клике на кнопку происходит изменение контента 2`] = `
<body>
  <div>
    <ul
      class="tabs-nav"
    >
      <li
        class="tabs-nav__item"
      >
        <button
          class="tabs-nav__btn"
        >
          Item 1
        </button>
      </li>
      <li
        class="tabs-nav__item active"
      >
        <button
          class="tabs-nav__btn"
        >
          Item 2
        </button>
      </li>
    </ul>
    <div
      class="tabs-content"
      data-testid="tabs"
    >
      Children 2
    </div>
  </div>
</body>
`;
